### PR TITLE
issues/158/examples for working api with javascript frontend

### LINKED ISSUE #158: Gorilla CSRF always sets the same cookie, even as csrf.Token(r) varies
I cannot get any version of the documentation working with Javascript.
I run a get with Axios to set the header on the axios instance, then I post with the header and cookie included.
The problem is that the csrf cookie that Gorilla sets is always the same cookie, every single time, across restarts, and browsers, and curl, and Postman, it is always `_gorilla_csrf=MTYzNzEwOTA5NHxJbkYwVGk5VFpVcFVWQ3RYV0ZOeU0yWjFRa3d5T0RoTVUycGtjbkZhU2pGd1p6ZDVNamxRWVcxV2FFRTlJZ289fLeqCKHNIBazzPu_ILjc2dkBBbeqBjy52MgsxkeVOvfg; HttpOnly; SameSite=Lax`

Therefore I can never successfully POST, because my header (always regenerated and fresh) and cookie (always the same) never match.

Here is my code, stripped down as much as possible. I use Chi router because it is what I am familiar with, but that difference from the docs should not matter here.

Go (`go version go1.17.2 darwin/amd64`):
```go
package main

import (
	"fmt"
	"log"
	"net/http"
	"time"

	"github.com/go-chi/chi"
	"github.com/go-chi/chi/middleware"
	"github.com/gorilla/csrf"
	"github.com/rs/cors"
)

func main() {
	router := chi.NewRouter()

	// Suggested basic middleware stack from chi's docs
	router.Use(middleware.RequestID)
	router.Use(middleware.RealIP)
	router.Use(middleware.Logger)
	router.Use(middleware.Recoverer)

	CSRF := csrf.Protect(
		[]byte("place-your-32-byte-long-key-here"),
		csrf.RequestHeader("X-CSRF-Token"),
		csrf.Secure(false),
		csrf.TrustedOrigins([]string{"*"}),
		csrf.MaxAge(0),
	)

	// Routing to API handlers
	router.Route("/api", func(router chi.Router) {
		router.With(CSRF).Get("/", Get)
		router.With(CSRF).Post("/", Post)
	})

	corsRouter := cors.New(cors.Options{
		AllowCredentials: true,
		AllowedHeaders:   []string{"Content-Type", "X-CSRF-Token"},
		ExposedHeaders:   []string{"Content-Type", "X-CSRF-Token"},
		AllowedMethods:   []string{"GET", "POST", "OPTIONS"},
		AllowedOrigins:   []string{"*"},
		Debug:            true,
	}).Handler(router)

	srv := &http.Server{
		Handler:      corsRouter,
		Addr:         "localhost" + ":" + "8000",
		ReadTimeout:  time.Duration(15) * time.Second,
		WriteTimeout: time.Duration(10) * time.Second,
		IdleTimeout:  time.Duration(5) * time.Second,
	}

	fmt.Printf("starting http server on port %s...\n", "8000")
	log.Fatal(srv.ListenAndServe())

}


func Get(w http.ResponseWriter, r *http.Request) {
	w.Header().Add("X-CSRF-Token", csrf.Token(r))
}

func Post(w http.ResponseWriter, r *http.Request) {}
```

Javascript:
```javascript
let get = async() => {
    let csrfToken = ""
    try {
        let resp = await axios.get("http://localhost:8000/api/")
        console.log(resp)
        console.log(resp.headers)
        csrfToken = resp.headers["x-csrf-token"]
        console.log(csrfToken)
    } catch (err) {
        // Handle the exception
        console.log(err)
    }

    const instance = axios.create({
        headers: {"X-CSRF-Token": csrfToken}
    })
    console.log(instance)
    return instance
}

let post = async(instance) => {
    try {
        let resp = await instance.post("http://localhost:8000/api/", {})
        console.log(resp)
    } catch (err) {
        console.log(err)
    }
}

get()
    .then(axiosInstance => {
        post(axiosInstance)
    })
```

### PR DESCRIPTION
**Pull request by francoposa**
Fixes #158, which is essentially that
1. none of the examples in the README for working with a JavaScript frontend will work without proper CORS config on the backend
2. there is no example at all for using the HTTP header instead of getting the CSRF token from the hidden form field

**Summary of Changes**

I have merged/copied over these simplified examples from my own repository of working examples.

I was not sure how the maintainers may want to reference these examples in the main README. Copying them over to the README verbatim would be putting a lot of code into the README, but without changing the current README, the content there differs significantly from the examples.

### CODE REVIEW

File: examples/api-backends/gorilla-mux/main.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+func main() {
+	router := mux.NewRouter()
+
+	loggingMiddleware := func(h http.Handler) http.Handler {
+		return handlers.LoggingHandler(os.Stdout, h)
+	}
+	router.Use(loggingMiddleware)
+
+	CSRFMiddleware := csrf.Protect(
+		[]byte("place-your-32-byte-long-key-here"),
```
Comment (robbat2) [inline examples/api-backends/gorilla-mux/main.go:25]: hoist the literal value out of the code, into a constant?

Comment (francoposa) [reply examples/api-backends/gorilla-mux/main.go:25]: definitely open to this; in my own csrf examples repo I have everything factored out to be configurable through cli flags and config files.

For these I tried to follow the style of the existing docs and keep it as short as possible.
It seems like in general, example/quickstart docs tend to opt for hardcoding

**Comment by francoposa** [comment]
@DavidLarsKetch do you have any input on how to approach the README?

**Comment by coreydaley** [comment]
Would you mind updating your example to use go1.20 and update the dependencies in your go.mod? Then I think we can get this merged.

**Review by coreydaley — APPROVED** [review]

**Comment by canrozanes** [comment]
Hi @francoposa
> none of the examples in the README for working with a JavaScript frontend will work without proper CORS config on the backend

Could you point me to a resource about this? Is CORS a pre-requisite for CSRF protection? If I setup gorilla/csrf on my SPA without CORS, will my site still be vulnerable to CSRF?

**Pull request merged**